pkg/identities/service.go

```diff
@@ -240,3 +249,514 @@ func NewService(kratos kClient.IdentityAPI, authz AuthorizerInterface, tracer tr
+func (s *V1Service) GetIdentityEntitlements(ctx context.Context, identityId string, params *resources.GetIdentitiesItemEntitlementsParams) (*resources.PaginatedResponse[resources.EntityEntitlement], error) {
+	ctx, span := s.core.tracer.Start(ctx, "identities.V1Service.GetIdentityEntitlements")
+	defer span.End()
+
+	paginator := types.NewTokenPaginator(s.core.tracer, s.core.logger)
+
+	nextToken := ""
+
+	if params != nil && params.NextPageToken != nil {
+		nextToken = *params.NextPageToken
+	}
+
+	if err := paginator.LoadFromString(ctx, nextToken); err != nil {
+		s.core.logger.Error(err)
+	}
+
+	permissions, pageTokens, err := s.openfgaStore.ListPermissions(ctx, fmt.Sprintf("user:%s", identityId), paginator.GetAllTokens(ctx))
+
+	if err != nil {
+		return nil, v1.NewUnknownError(err.Error())
+	}
+
+	paginator.SetTokens(ctx, pageTokens)
+	metaParam, err := paginator.PaginationHeader(ctx)
+	if err != nil {
+		s.core.logger.Errorf("error producing pagination meta param: %s", err)
+		metaParam = ""
+	}
+
+	r := new(resources.PaginatedResponse[resources.EntityEntitlement])
+	r.Meta = resources.ResponseMeta{Size: len(permissions)}
+	r.Data = make([]resources.EntityEntitlement, 0)
+	r.Next.PageToken = &metaParam
+
+	for _, permission := range permissions {
+
+		entity := strings.SplitN(permission.Object, ":", 2)
+		r.Data = append(
+			r.Data,
+			resources.EntityEntitlement{
+				Entitlement: permission.Relation,
+				EntityType:  entity[0],
+				EntityId:    entity[1],
+			},
+		)
+	}
+
+	return r, nil
+}
+
+// PatchIdentityEntitlements performs addition or removal of an Entitlement to/from an Identity.
+func (s *V1Service) PatchIdentityEntitlements(ctx context.Context, identityId string, entitlementPatches []resources.IdentityEntitlementsPatchItem) (bool, error) {
+	ctx, span := s.core.tracer.Start(ctx, "identities.V1Service.PatchIdentityEntitlements")
+	defer span.End()
+
+	additions := make([]ofga.Permission, 0)
+	removals := make([]ofga.Permission, 0)
+	for _, p := range entitlementPatches {
+		permission := ofga.Permission{
+			Relation: p.Entitlement.Entitlement,
+			Object:   fmt.Sprintf("%s:%s", p.Entitlement.EntityType, p.Entitlement.EntityId),
+		}
+
+		if p.Op == "add" {
+			additions = append(additions, permission)
+		} else if p.Op == "remove" {
+			removals = append(removals, permission)
+		}
+	}
+
+	if len(additions) > 0 {
+		err := s.openfgaStore.AssignPermissions(ctx, fmt.Sprintf("user:%s", identityId), additions...)
+
+		if err != nil {
+			return false, v1.NewUnknownError(err.Error())
+		}
+	}
+
+	if len(removals) > 0 {
+		err := s.openfgaStore.UnassignPermissions(ctx, fmt.Sprintf("user:%s", identityId), removals...)
+		if err != nil {
+			return false, v1.NewUnknownError(err.Error())
+		}
+	}
+
+	return true, nil
+}
+
+type Config struct {
+	Name         string
+	Namespace    string
+	K8s          coreV1.CoreV1Interface
+	OpenFGAStore OpenFGAStoreInterface
```

Comment:
Just a **NONBLOCKING** question: why `Config` has a field `OpenFGAStore`? Seems that `Config` is closer to k8s configmap stuff.

Reply:
it follows the same pattern as https://github.com/canonical/identity-platform-admin-ui/blob/main/pkg/schemas/service.go#L27-L32

ideally we never have more than 4/5 args in a constructor and use a config type